app/app.go

```diff
@@ -576,15 +611,17 @@ func NewStrideApp(
 
 	// Stack two contains
 	// - IBC
+	// - ratelimit
```

Comment:
What's the thinking around the ordering here? I'm not sure it matters, just curious. Conceptually, I'd probably move this down closer to transfer (because it's lower level), but not sure there's actually a reason to do so!

Reply:
Also, if we change this, we probably need to update this LOC https://github.com/notional-labs/stride/blob/acd8c57188652ce42c53a09c82a2d5bbe49b83af/x/icacallbacks/keeper/keeper.go#L115

Reply:
No specific reasoning. I'm fine with moving it closer to transfer

Reply:
On second thought, I actually think the most logical ordering of everything is probably:

IBC
ratelimit
transfer
records

At the end of the day, it's pretty arbitrary since traffic floats in both directions. But records feels closest to the core application and rate limit/transfer feel more like related to the actually transport/packets.

Reply:
That framework makes sense. transfer seems closer to IBC than ratelimit (you can have transfer w/o ratelimit, but not vice-versa)

Reply:
I'm not sure I follow the with/without argument? 

I think of transfer as closer to the application (e.g. it's what mints/burns tokens to users) and ratelimit more as middleware that is meant to sit in between the relaying of packets (IBC) and the handling of those packets (Base App)

Reply:
Another way we might want to think about it, is I think depending on the ordering, we might have to revert the action of the previous module.

---

So if we have 
IBC
transfer
ratelimit

Then for a receive packet (travels down the stack), we have to burn/re-escrow any tokens that were minted/un-escrowed by transfer.

---

Similarly if we have

IBC
ratelimit
transfer

Then for a send packet (travels up the stack), we have to mint/un-escrow any tokens that were burned/escrowed by transfer

Reply:
> the relaying of packets (IBC) and the handling of those packets

I was holding we'd only ever rate limit transfer packets, so it would only make sense to include ratelimit in a middleware stack that also includes transfer. I guess it's possible we rate-limit other types of packets as well! Putting it next to IBC makes sense

> Another way we might want to think about it, is I think depending on the ordering, we might have to revert the action of the previous module.

Would we want the entire middleware stack to atomically succeed/fail? Looks like right now `SendRateLimitedPacket` throws an error (which should revert all other calls in the middleware stack), but `ReceiveRateLimitedPacket` returns a `channeltypes.NewErrorAcknowledgement(err)` (which I don't think will revert logic executed by other middlewares, but not 100% sure)

Reply:
yeah I think right now there's a bug - I'm refactoring the middleware wiring as we speak.

I think we need to revert one of the cases, and it depends on the ordering of the stack (as described in my last example), but right now we're reverting neither.

I'm also wondering if it makes sense to just have separate stacks?

Reply:
Does [this error being thrown](https://github.com/Stride-Labs/stride/pull/556/files#diff-62a2ad25069547e2e6afd9aaa01547c698b11f7725c70ef4ce4c5d2c4080e83dR30) not revert sends? I would have expected the error to propagate through the middleware stack

> I'm also wondering if it makes sense to just have separate stacks?

Can you elaborate on this?

Reply:
Hmm actually yeah I think you're right with the current way we set it up. It think the SendPacket you linked above is called in ibc-go [here](https://github.com/cosmos/ibc-go/blob/main/modules/apps/transfer/keeper/relay.go#L136)

But that's just cause we added the rateLimitKeeper as the ics4Wrapper in the transfer keeper - but it's not clear to me yet if that's the right way to set it up

Reply:
Revisited this - 

I think transfer has to sit at the bottom of the stack because the transfer `IBCModule` does not have the `app` attribute used for middleware (e.g. [this](https://github.com/cosmos/ibc-go/blob/v5.1.0/modules/apps/transfer/ibc_module.go#L21-L23) vs [this](https://github.com/cosmos/ibc-go/blob/v5.1.0/modules/apps/29-fee/ibc_middleware.go#L21-L23)). Meaning, if you had transfer at the top of the stack, then there's no way for the Transfer `OnRecvPacket` callback to pass the packet down to a different middleware.

---

That then brings us full circle back to the question of whether the stack should be
```
records
ratelimit
transfer

OR

ratelimit
records
transfer
```

And between those two, I think I'd side with your OG suggestion that keeping ratelimit closest to transfer probably makes the most sense. 

Sorry for the runaround!

Reply:
And then back to the error question…

(once again for context)
* SendPacket flows `transfer` → `ratelimit` → `records`
* RecvPacket flow `records` → `ratelimit` → `transfer`

<br>
There are the following two error cases in the rate limit module:
* SendPacket - As implemented already and discussed above, I think throwing an error from `ratelimit` is adequate because it will propagate back to the transfer module and should atomically fail & discard anything the transfer module had done already
* OnRecvPacket - As implemented, I think returning an AckError is adequate here because nothing has been done yet on the Stride side of things that would need to get undone (contrary to something like OnAckPacket). And when we return the ack error, it should propagate back to `records` and then get emitted in the core IBC keeper